### PR TITLE
Ewlumpkin/issue2

### PR DESCRIPTION
**Pull request by ewlumpkin**
Addressed: 
https://github.com/cooperlees/json2netns/pull/10#discussion_r694266375
https://github.com/cooperlees/json2netns/pull/10#discussion_r694266838
https://github.com/cooperlees/json2netns/pull/10#discussion_r694269698
https://github.com/cooperlees/json2netns/pull/10#discussion_r694276191

Comments:
https://github.com/cooperlees/json2netns/pull/10#pullrequestreview-736511089

> Possible future improvements.
> 
> * Specific route based tests
>   
>   * e.g. showing validation works

* Added validation testing and included coverage in setup.py
   * Still need some additional tests for functionality e.g. subprocess mock(s) for route_add

> * Swap the "route/attributes" dict for properties of the "Route" object or make a RouteAttributes named tuple OR dataclass
>   
>   * Why? Stricter typing than a dict

* Changed Route class to dataclass and added typing to attributes

> * From a JSON config we can use **kwargs to construct the objects nicely

* Had some guesses about implementation but wanted to see your idea before changing it.

Open to any and all suggestions

### CODE REVIEW

File: src/json2netns/route.py [hunk 1]
```diff
@@ -1,139 +1,146 @@
...
 
-    def __proto_match_validated(self, route_test: Dict) -> bool:
+    # TODO Add support for IPv4 via IPv6 next hops (should probably open separate issue)
+    def __proto_match_validated(self) -> bool:
         """Check to make sure the destination IP and next-hop match protocols (v4/v6)"""
-        if (
-            ip_network(route_test["dest_prefix"]).version
-            == ip_address(route_test["next_hop_ip"]).version
-        ):
+        if ip_network(self.dest_prefix).version == ip_address(self.next_hop_ip).version:
```
Comment (cooperlees) [inline src/json2netns/route.py:26]: So much cleaner!

File: src/json2netns/tests/netns.py [hunk 2]
```diff
@@ -62,6 +62,9 @@ def test_delete(self) -> None:
             self.test_ns.delete()
             self.assertEqual(1, mock_run.call_count)
 
+    def test_route_add(self) -> None:
+        pass
```
Comment (cooperlees) [inline src/json2netns/tests/netns.py:66]: Forgot to do?

File: src/json2netns/tests/route.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    """Useful function to use with patch() to see what's being ran"""
+    LOG.debug(f"CMD ARGS: {args} - {kwargs}")
+    # Always return 0 - Naturally return non 0 for errors
+    return CompletedProcess(args, 0)
+
+
+class RouteTests(unittest.TestCase):
+    def setUp(self) -> None:
+        self.config = Config(SAMPLE_JSON_CONF_PATH).load()
+        print(self.config)
```
Comment (cooperlees) [inline src/json2netns/tests/route.py:30]:
```suggestion
        self.config = Config(SAMPLE_JSON_CONF_PATH).load()
```
Debug to remove?

File: src/json2netns/tests/route.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+                # Initialize route obj
+                route_obj = Route(
+                    route_name,
+                    ns_name,
+                    route["dest_prefix"],
+                    route["next_hop_ip"],
+                    route["egress_if_name"],
+                )
+                # create a list of known good route objects to use
+                self.route_list.append(route_obj)
```
Comment (cooperlees) [inline src/json2netns/tests/route.py:47]:
```suggestion
                # create a list of known good route objects to use for testing
                self.route_list.append(route_obj)
```

File: src/json2netns/tests/route.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+            "bad_nexthop_route", "bad_ns", "192.168.1.0/32", "192.168.1.6969", "eth69"
+        )
+
+    def test_proto_match_validated(self) -> None:
+        # validate protocols match
+        for route in self.route_list:
+            self.assertIs(
+                route._Route__proto_match_validated(),
+                True,
+            )
```
Comment (cooperlees) [inline src/json2netns/tests/route.py:67]:
```suggestion
            self.assertTrue(
                route._Route__proto_match_validated()
            )
```
FYI - This exists

File: src/json2netns/tests/route.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+        # validate protocols don't match
+        self.assertIs(
+            self.mismatch_route_obj._Route__proto_match_validated(),
+            False,
+        )
```
Comment (cooperlees) [inline src/json2netns/tests/route.py:73]:
```suggestion
        self.assertFalse(
            self.mismatch_route_obj._Route__proto_match_validated(),
        )
```
Likewise

**Review by cooperlees — CHANGES_REQUESTED** [review]
Love the use of data class - I will prob change some of the other classes to be a dataclass

- I offered some NITs - Nothing big

Do you intend to finish the other unitests with `pass`?

**Comment by ewlumpkin** [comment]
> Love the use of data class - I will prob change some of the other classes to be a dataclass
> 
> * I offered some NITs - Nothing big
> 
> Do you intend to finish the other unitests with `pass`?

Awesome, thanks.

Still wrapping my head around patch and mock.  Not really sure how to test those remaining methods at the moment.

File: setup.py [hunk 7]
```diff
@@ -16,8 +16,8 @@
         "json2netns/consts.py": 100,
         "json2netns/interfaces.py": 90,
         "json2netns/main.py": 70,
-        "json2netns/netns.py": 71,
-        "json2netns/route.py": 47,
+        "json2netns/netns.py": 76,
+        "json2netns/route.py": 76,
```
Comment (cooperlees) [inline setup.py:20]: Damn - nice!

File: src/json2netns/tests/netns.py [hunk 8]
```diff
@@ -63,7 +65,13 @@ def test_delete(self) -> None:
             self.assertEqual(1, mock_run.call_count)
 
     def test_route_add(self) -> None:
-        pass
+        with patch.object(Route, "get_route") as mock_get_route, patch.object(
```
Comment (cooperlees) [inline src/json2netns/tests/netns.py:68]: TIL - Never used this. This is nice.

File: src/json2netns/tests/route.py [hunk 9]
```diff
@@ -89,7 +80,34 @@ def test_route_validated(self) -> None:
...
+            self.assertTrue(self.route_list[1].route_exists())
 
     def test_get_route(self) -> None:
-        pass
+        # Use first host route in sample.json to test -> 10.6.9.6 via 10.1.1.2
+        # Checks within this method are done above, thus mocked
+        with patch.object(
+            Route, "_Route__proto_match_validated", return_value=True
+        ), patch.object(
+            Route, "_Route__route_validated", return_value=True
```
Comment (cooperlees) [inline src/json2netns/tests/route.py:106]: Dunders everywhere!

**Review by cooperlees — APPROVED** [review]
LGTM - Thanks!

**Pull request merged**